File: 5 - Sorting/SelectionSort.java
package com.proglang;

import java.util.Arrays;
import java.util.Scanner;

public class SelectionSort {
    public static void main(String[] args) {

        /*  Selection Sort
            Time Complexity -
                Best Case - Omega(n^2)
                Average Case - Theta(n^2)
                Worst Case - BigO(n^2)
         */

        Scanner scanner = new Scanner(System.in);
        System.out.print("n: ");
        int n = scanner.nextInt();

        scanner.nextLine();

        String[] s = scanner.nextLine().split(" ");
        int[] array = new int[n];

        for(int i=0; i<n; i++) {
            array[i] = Integer.parseInt(s[i]);
        }

        int tmp, min, minInd, swapCount=0;
        for (int i=0; i<n-1; i++) {
            min = array[i];
            minInd = i;

            for (int j=i+1; j<n; j++) {
                if (min > array[j]) {
                    min = array[j];
                    minInd = j;
                }
            }

            if(minInd!=i) {
                tmp = array[i];
                array[i] = min;
                array[minInd] = tmp;
                swapCount++;
                System.out.println("swap " + swapCount + " : " + Arrays.toString(array));
            }
        }

        System.out.println("Sorted Array : " + Arrays.toString(array));
        System.out.println("swapCount : " + swapCount);
    }
}

File: DSA - Python/Sorting/selection_sort.py
# Selection sort
# O(n^2) time | O(1) space
def selection_sort(array):
    n = len(array)
    for k in range(0, n):
        pos = smallest(array, k, n)
        array[k], array[pos] = array[pos], array[k]
    return array


def smallest(array, k, n):
    pos = k
    small = array[k]
    for i in range(k+1, n):
        if array[i] < small:
            small = array[i]
            pos = i
    return pos


def selectionSort(array):
    currentIdx = 0
    while currentIdx < len(array) - 1:
        smallestIdx = currentIdx
        for i in range(currentIdx + 1, len(array)):
            if array[smallestIdx] > array[i]:
                smallestIdx = i
        swap(currentIdx, smallestIdx, array)
        currentIdx += 1
    return array


def swap(i, j, array):
    array[i], array[j] = array[j], array[i]


arr = list(map(int, input().split(' ')))
print(selection_sort(arr))
print(selectionSort(arr))
